File: solutions/fsharp/robot-simulator/1/RobotSimulator.fs
module RobotSimulator

type Direction = North | East | South | West
type Position = int * int

let create direction position =
    (direction, position)

let rec move instructions robot =

    let (|Instruction|_|) p s =
        if s <> "" && s[0] = p then
            Some(s[1..])
        else
            None

    let (direction, position) = robot
    let (x, y) = position

    match (instructions, direction) with

    // Turn left
    | (Instruction 'L' next, North) -> move next (West, position)
    | (Instruction 'L' next, East)  -> move next (North, position)
    | (Instruction 'L' next, South) -> move next (East, position)
    | (Instruction 'L' next, West)  -> move next (South, position)

    // Turn right
    | (Instruction 'R' next, North) -> move next (East, position)
    | (Instruction 'R' next, East)  -> move next (South, position)
    | (Instruction 'R' next, South) -> move next (West, position)
    | (Instruction 'R' next, West)  -> move next (North, position)

    // Advance
    | (Instruction 'A' next, North) -> move next (direction, (x, y + 1))
    | (Instruction 'A' next, East)  -> move next (direction, (x + 1, y))
    | (Instruction 'A' next, South) -> move next (direction, (x, y - 1))
    | (Instruction 'A' next, West)  -> move next (direction, (x - 1, y))

    // Arrived - out of instructions
    | ("", _) -> robot

    | _ -> failwith "Invalid instruction"

File: solutions/typescript/robot-simulator/1/robot-simulator.ts
export class InvalidInputError extends Error {
  constructor(message: string) {
    super()
    this.message = message || 'Invalid Input'
  }
}

const Directions = ['north', 'east', 'south', 'west']
const Advances = [[0, 1], [1, 0], [0, -1], [-1, 0]]
type Direction = typeof Directions[number]
type Coordinates = [number, number]

export class Robot {
  direction: Direction
  position: Coordinates
  
  constructor() {
    this.direction = 'north'
    this.position = [0, 0]
  }
  
  get bearing(): Direction {
    return this.direction
  }

  get coordinates(): Coordinates {
    return this.position
  }

  place(params: { x: number; y: number; direction: string }) {
    if (Directions.includes(params.direction)) {
      this.direction = params.direction as Direction
    } else {
      throw new InvalidInputError(params.direction)
    }
    this.position = [params.x, params.y]
  }

  evaluate(instructions: string) {
    for (let step of instructions) {
      switch(step) {
        case 'R':
          this.turnRight()
          break
        case 'L':
          this.turnLeft()
          break
        case 'A':
          this.advance()
          break
      }
    }
  }

  turnRight() {
    let dirIdx = Directions.indexOf(this.direction) + 1
    this.direction = Directions[dirIdx % Directions.length]
  }

  turnLeft() {
    let dirIdx = Directions.indexOf(this.direction) + Directions.length - 1
    this.direction = Directions[dirIdx % Directions.length]
  }

  advance() {
    let dirIdx = Directions.indexOf(this.direction)
    let [posX, posY] = this.position
    let [movX, movY] = Advances[dirIdx]

    this.position = [posX + movX, posY + movY]
  }
}
